util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+
+    for k in range(num_words):
+        # Assemble native OTP word and uniquify annotation for comments
+        word = 0
+        word_ann = OrderedDict()
+        for j in range(bytes_per_word):
+            idx = k * bytes_per_word + j
+            word += data[idx] << (j * 8)
+            if annotation[idx] not in word_ann:
+                word_ann.update({annotation[idx]: "1"})
+
+        # This prints the byte offset of the corresponding
+        # payload data in the memory map (excluding ECC bits)
+        word_ann = ' // {:06x}: '.format(k * bytes_per_word) + ', '.join(
+            word_ann.keys())
+
+        # ECC encode
+        word_bin = format(word, bin_format_str)
+        word_bin = ecc_encode(config, word_bin)
+        word_hex = format(int(word_bin, 2), hex_format_str)
+        memory_words += word_hex + word_ann + '\n'
+
+    log.info('Done.')
+
+    return memory_words
+
+
+class OtpMemImg(OtpMemMap):
+
+    # LC state object
+    lc_state = []
+
+    def __init__(self, lc_state_config, otp_mmap_config, img_config):
+
+        # Initialize memory map
+        super().__init__(otp_mmap_config)
+
+        # Initialize the LC state and OTP memory map objects first, since
+        # validation and image generation depends on them
+        self.lc_state = LcStEnc(lc_state_config)
+
+        # Validate memory image configuration
+        log.info('')
+        log.info('Parse OTP image specification.')
+
+        # Encryption sanity test with known test vector
+        enc_test = _present_64bit_encrypt(
+            int("0123456789abcdef", 16),
+            int("0123456789abcdef0123456789abcdef", 16))
+
+        if enc_test == int('0e9d28685e671dd6', 16):
+            log.info('Encryption module test OK')
+        else:
+            raise RuntimeError('Encryption module test failed')
+        log.info('')
+
+        if 'seed' not in img_config:
+            raise RuntimeError('Missing seed in configuration.')
+
+        img_config['seed'] = check_int(img_config['seed'])
+        log.info('Seed: {0:x}'.format(img_config['seed']))
+        log.info('')
+
+        # Re-initialize with seed to make results reproducible.
+        random.seed(OTP_IMG_SEED_DIVERSIFIER + int(img_config['seed']))
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Adding values to {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+        log.info('')
+        log.info('Parsing OTP image successfully completed.')
+
+    def merge_part_data(self, part):
+        '''This validates and merges the partition data into the memory map dict'''
+
+        part.setdefault('lock', 'false')
+        part['lock'] = check_bool(part['lock'])
+        part.setdefault('items', [])
```

Comment:
My comment above about checking for unexpected keys is probably more important here. Because e.g. the `lock` key is optional, if you don't do the check then you'll silently ignore a field of a config with a typo that sets `lokk`, resulting in an unlocked partition.

I'd consider defining a class to represent partitions. Then you can do all the parsing / validation in its constructor and just do "merging" here. (To be really cute, the `LIFE_CYCLE` partition could have its own subclass)

Reply:
Jep, I've done added key account here as well.

I have thought about making classes for the item and partition objects.
But to make this clean we would also have to do that in the base class, which means that this has to bubble through all the templates that already exist.
I am probably going to defer that refactor pass to another PR.

Reply:
Seems wise!